aip/client-libraries/4234.md

```diff
@@ -0,0 +1,110 @@
+- `google.iam.v1.IAMPolicy`
+- `google.longrunning.Operations`
+
+**Note:** The list of supported mixin services and the mixin services
+themselves change infrequently. Any new common service to be supported should be
+added to the list here and generators should be updated.
+
+To be generated into a client library, a mixin service **must** be declared
+under [apis] in a [google.api.Service]. Furthermore, it **must** have
+[google.api.http] bindings declared for all of its RPCs in the [http]
+configuration of the same [google.api.Service] (more details on these bindings
+in AIP-127).
+
+
+## Implementing mixin support
+
+### Generator configuration
+
+Client library generators **must** accept the file path of a
+[google.api.Service] in YAML form. This flag **must** be optional - meaning
+the flag **must not** be required for basic client generation that omits the
+mixin methods.
+
+### Mixin API client configuration
+
+If a mixin API utilizes client library configuration such as
+`google.api.method_signature` or request header injection, generators **may**
+support it in the generated mixin methods if the configuration is accessible.
+
+Where client library configuration calls for a fully-qualified name, the
+_mixin's_ fully-qualified name **must** be used. That is to say, the host
+service **must not** refer to a mixin element as if it was in its own package.
+
+### Generating mixin methods
+
+The mixin API RPCs **should** be generated as methods on the surface of the
+host API's service client library, alongside the host service's RPCs. This
+presents them as top-level methods on the client, but under-the-hood, the
+appropriate mixin API stub or request should be invoked.
+
+**Note:** For gRPC clients, the mixin API gRPC stub **must** be used in order
+to properly construct the gRPC request.
+
+### Multiple host service clients
+
+If there are multiple services defined by the host API that would result in
+multiple clients being generated, generators **must** include the mixin
+methods on all eligible, generated service clients.
+
+### Overriding a duplicate RPC
+
+Client library generators **must not** generate a mixin method on any host
+service client if a host service already defines an RPC with the same name. For
+example, take the following service definition:
+
+```proto
+service LibraryService {
+  rpc ListBooks(ListBookRequest) returns (ListBooksResponse);
+
+  // Other host API RPCs...
+
+  // Redefinitions of the google.iam.v1.IAMPolicy mixin service.
+  rpc GetIamPolicy(google.iam.v1.GetIamPolicyRequest)
+      returns (google.iam.v1.Policy);
+
+  rpc SetIamPolicy(google.iam.v1.SetIamPolicyRequest)
+      returns (google.iam.v1.Policy);
+
+  rpc TestIamPermissions(google.iam.v1.TestIamPermissionsRequest)
+      returns (google.iam.v1.TestIamPermissionsResponse);
+}
+```
+
+Should the host API declare the `google.iam.v1.IAMPolicy` as a mixin service,
+client library generators **must not** generate the resulting mixin methods for
```

Comment:
s/`the resulting mixin methods`/`the resulting mixin methods that match the names explicitly declared above, but they **must** generate the other methods for this mixin`/ right?

Reply:
Hm interesting question. Since `IAMPolicy` is the only API we have that is redeclared by each individual API, we only really have the one use case to go off of. With `IAMPolicy` every RPC is redeclared, always, so omission of some RPCs but not others hasn't come up before. This is a question worth answering to make things future proof. Your interpretation certainly would work for this one use case we have today, because every RPC is always redeclared, I just don't know if that would be right for another such API. WDYT?